src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,218 @@ namespace NServiceBus
-                    lockers.TryRemove(value.LockTokenKey, out lockToken);
+                    throw new Exception("Saga can't be completed as it was updated by another process.");
+                }
+
+                // saga removed
+                // clean the index
+                if (Equals(entry.CorrelationId, NoCorrelationId) == false)
+                {
+                    byCorrelationId.TryRemoveConditionally(entry.CorrelationId, sagaData.Id);
                 }
             });
-            return TaskEx.CompletedTask;
+
+            return Task.FromResult(0);
         }
 
-        public Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
+        public Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context)
+            where TSagaData : IContainSagaData
         {
-            Guard.AgainstNull(nameof(propertyValue), propertyValue);
+            Entry value;
 
-            foreach (var entity in data.Values)
+            if (sagas.TryGetValue(sagaId, out value))
             {
-                if (!(entity.SagaData is TSagaData))
-                {
-                    continue;
-                }
+                context.Set(sagaId.ToString(), value);
 
-                var prop = typeof(TSagaData).GetProperty(propertyName);
-                if (prop == null)
-                {
-                    continue;
-                }
-                var existingValue = prop.GetValue(entity.SagaData);
+                var data = value.GetSagaCopy();
 
-                if (existingValue.ToString() != propertyValue.ToString())
+                if (data is TSagaData)
                 {
-                    continue;
+                    return Task.FromResult((TSagaData)data);
                 }
-                var sagaData = entity.Read<TSagaData>();
-                return Task.FromResult(sagaData);
             }
+
             return Task.FromResult(default(TSagaData));
         }
 
-        public Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
+        public Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
         {
-            VersionedSagaEntity result;
-            if (data.TryGetValue(sagaId, out result) && result?.SagaData is TSagaData)
+            var key = GetKey(typeof(TSagaData), propertyName, propertyValue);
+            Guid id;
+
+            if (byCorrelationId.TryGetValue(key, out id))
             {
-                var sagaData = result.Read<TSagaData>();
-                return Task.FromResult(sagaData);
+                // this isn't updated atomically and may return null for an entry that has been indexed but not inserted yet
```

Comment:
fair point. In general I think we need to be very careful with making assumptions about retry behaviors as they could easily be disabled. But on the saga persistence level, there is always a risk of race conditions and therefore concurrency exceptions, so I think I can withdraw my concern here. 👍